src/gc.c

```diff
@@ -539,16 +540,28 @@ static void run_finalizers(void)
     arraylist_free(&copied_list);
 }
 
-void jl_gc_inhibit_finalizers(int state)
+JL_DLLEXPORT int jl_finalizers_enable(int on)
 {
-    // NOTE: currently only called with the codegen lock held, but might need
-    // more synchronization in the future
-    if (jl_gc_finalizers_inhibited && !state && !jl_in_finalizer) {
-        jl_in_finalizer = 1;
-        run_finalizers();
-        jl_in_finalizer = 0;
+    jl_tls_states_t *ptls = jl_get_ptls_states();
+    int prev = !ptls->disable_finalizers;
+    ptls->disable_finalizers = (on == 0);
+    if (on && !prev) {
+        // disable -> enable
+        // TODO: the atomics are in the wrong place / missing some thread locks
+        jl_atomic_fetch_add(&jl_finalizers_disabled_counter, -1);
+        if (jl_finalizers_disabled_counter == 0) {
+            int8_t was_in_finalizer = jl_in_finalizer;
+            jl_in_finalizer = 1;
+            run_finalizers();
+            jl_in_finalizer = was_in_finalizer;
+        }
     }
-    jl_gc_finalizers_inhibited = state;
+    else if (prev && !on) {
+        // enable -> disable
+        jl_atomic_fetch_add(&jl_finalizers_disabled_counter, 1);
+        // TODO: check if finalizers are running and wait for them to finish
```

Comment:
Whether we need a Dict lock or not depends on what's the thread safety guarantee we provide for stdlib. Currently, the user always has to synchronize access to a dict from multiple threads and there's nothing special about whether it is running from within a finalizer or not and the only feature Base has to provide is that finalizer running on the same thread will not corrupt the data structure.

If, however, we begin to guarantee thread safety of Base data structures it would indeed be better to use that for finalizer synchronization too.